packages/datadog-plugin-rhea/src/index.js

```diff
@@ -0,0 +1,210 @@
+      const { host, port } = getHostAndPort(this.connection)
+      return tracer.trace('amqp.send', {
+        tags: {
+          'component': 'rhea',
+          'resource.name': name,
+          'service.name': config.service || `${tracer._service}-amqp-producer`,
+          'span.kind': 'producer',
+          'amqp.link.target.address': name,
+          'amqp.link.role': 'sender',
+          'out.host': host,
+          'out.port': port
+        }
+      }, (span, done) => {
+        addDeliveryAnnotations(msg, tracer, span)
+        const delivery = send.apply(this, arguments)
+        delivery[dd] = { done, span }
+        return delivery
+      })
+    }
+  }
+}
+
+function createWrapConnectionDispatch (tracer, config) {
+  return function wrapDispatch (dispatch) {
+    return function dispatchWithTrace (eventName, obj) {
+      if (eventName === 'disconnected' && this.local_channel_map) {
+        const error = obj.error || this.saved_error
+        for (const key in this.local_channel_map) {
+          const session = this.local_channel_map[key]
+          if (session) {
+            finishDeliverySpans(session.incoming, error)
+            finishDeliverySpans(session.outgoing, error)
+          }
+        }
+      }
+      return dispatch.apply(this, arguments)
+    }
+  }
+}
+
+function createWrapReceiverDispatch (tracer, config, instrumenter) {
+  return function wrapDispatch (dispatch) {
+    return function dispatchWithTrace (eventName, msgObj) {
+      patchCircularBuffer(this, instrumenter)
+      if (eventName === 'message') {
+        const name = msgObj.receiver.options.source && msgObj.receiver.options.source.address
+          ? msgObj.receiver.options.source.address : 'amq.topic'
+        const childOf = tracer.extract('text_map', msgObj.message.delivery_annotations)
+        return tracer.trace('amqp.receive', {
+          tags: {
+            'component': 'rhea',
+            'resource.name': name,
+            'service.name': config.service || tracer._service,
+            'span.kind': 'consumer',
+            'amqp.link.source.address': name,
+            'amqp.link.role': 'receiver'
+          },
+          childOf
+        }, (span, done) => {
+          msgObj.delivery[dd] = { done, span }
+          msgObj.delivery.update = wrapDeliveryUpdate(msgObj.delivery.update)
+          return dispatch.apply(this, arguments)
+        })
+      }
+
+      return dispatch.apply(this, arguments)
+    }
+  }
+}
+
+function createWrapCircularBufferPopIf () {
+  return function wrapCircularBufferPopIf (popIf) {
+    return function wrappedPopIf (fn) {
+      const wrappedFn = entry => {
+        const shouldPop = fn(entry)
+        if (shouldPop && entry[dd]) {
+          const remoteState = entry.remote_state
+          const state = remoteState && remoteState.constructor && remoteState.constructor.composite_type
+            ? entry.remote_state.constructor.composite_type : undefined
```

Comment:
Redundant since the result of `remoteState && remoteState.constructor && remoteState.constructor.composite_type` will be undefined anyway if not defined.